File: C++/soln-stack-problems/valid-parentheses.cpp
/**
 @author Farheen Bano
  
 Reference-
 https://leetcode.com/problems/valid-parentheses/
*/

class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for(char c:s){
            if(c=='(' || c=='[' || c=='{')
                st.push(c);
            else if(c==')' && !st.empty() && st.top()=='(')
                st.pop();
            else if(c==']' && !st.empty() && st.top()=='[')
                st.pop();
            else if(c=='}' && !st.empty() && st.top()=='{')
                st.pop();
            else
                return false;
        }
        return st.empty();
    }
};

File: Python/soln-stack-problems/valid-parentheses.py
'''
 @author Farheen Bano
  
 Reference-
 https://leetcode.com/problems/valid-parentheses/
'''
class Solution:
    def isValid(self, s: str) -> bool:
        stack=[]
        
        for ch in s:
            if stack and ch==')' and stack[-1]=='(':
                stack.pop()
        
            elif stack and ch==']' and stack[-1]=='[':
                stack.pop()
                
            elif stack and ch=='}' and stack[-1]=='{':
                stack.pop()
                
            else:
                stack.append(ch)
                
        return len(stack)==0
    